File: php/features.php
<?php
declare(strict_types=1);

// TelegramMailingService SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class TelegramMailingServiceFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new TelegramMailingServiceBaseFeature();
            case "ratelimit":
                return new TelegramMailingServiceRatelimitFeature();
            case "retry":
                return new TelegramMailingServiceRetryFeature();
            case "test":
                return new TelegramMailingServiceTestFeature();
            case "timeout":
                return new TelegramMailingServiceTimeoutFeature();
            default:
                return new TelegramMailingServiceBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: py/telegrammailingservice_sdk/features.py
# TelegramMailingService SDK feature factory

from telegrammailingservice_sdk.feature.base_feature import TelegramMailingServiceBaseFeature
from telegrammailingservice_sdk.feature.ratelimit_feature import TelegramMailingServiceRatelimitFeature
from telegrammailingservice_sdk.feature.retry_feature import TelegramMailingServiceRetryFeature
from telegrammailingservice_sdk.feature.test_feature import TelegramMailingServiceTestFeature
from telegrammailingservice_sdk.feature.timeout_feature import TelegramMailingServiceTimeoutFeature


_FEATURES = {
    "base": lambda: TelegramMailingServiceBaseFeature(),
    "ratelimit": lambda: TelegramMailingServiceRatelimitFeature(),
    "retry": lambda: TelegramMailingServiceRetryFeature(),
    "test": lambda: TelegramMailingServiceTestFeature(),
    "timeout": lambda: TelegramMailingServiceTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
